### PR TITLE
tests: add initial patchbay tests

### PR DESCRIPTION
**Pull request by Frando**
## Description

Adds a first round of tests using our new network simulation framework, [patchbay](https://github.com/n0-computer/patchbay). Patchbay uses Linux network namespaces to create isolated network topologies with routers, NATs, and link impairment. This lets us test actual holepunching and NAT traversal in realistic conditions, without needing any external infrastructure.

The test suite covers establishing a direct path in different scenarios:

- Holepunching: two devices behind destination-independent NATs connect via relay, then upgrade to direct.
- Uplink switching: a device changes its network uplink (v4 and v6) mid-connection and recovers a direct path.
- Interface changes: network interfaces are added or removed while connected
- Link outage recovery: a device goes offline and comes back, direct connection resumes
- Degraded links: increasing levels of impairment (latency, jitter, packet loss) are applied to either side of a connection and holepunching still succeeds

Some tests that iroh doesn't pass yet are included but marked `#[ignore]`. They should be un-ignored as we improve things.

### Running the tests

On Linux, things just work with user namespace support:
```sh
cargo make patchbay
# expands to:
cargo nextest run -p iroh --features qlog --test patchbay --profile patchbay

# run a specific test with logs:
RUST_LOG=trace cargo make patchbay holepunch_simple --nocapture
```

Test output is saved in `./target/testdir-current/patchbay/<test-name>`.  Patchbay currently collects tracing logs for each device, qlog files (if the feature is enabled), and endpoint metrics.
There's also a browser UI for viewing timelines, topologies, and logs. The UI can also be used to compare different test runs.
```sh
cargo install --git https://github.com/n0-computer/patchbay patchbay-cli
patchbay serve --testdir --open
```

On macOS you'll need to run the patchbay tests in a VM or container. The `patchbay` CLI includes a tool to set up a container or QEMU VM for this. 
```sh
# both commands default to the container backend on macOS and native on linux
patchbay test -p iroh --test patchbay
patchbay serve --testdir --open
# use --vm to force VM mode even on linux
```

See the [patchbay docs](https://n0-computer.github.io/patchbay/) and [README](https://github.com/n0-computer/patchbay) for more details.

## Notes and open questions

There's a draft PR with some more tests: #4065, but they need more thought and work to really cover what we want to test.

## Breaking Changes

None.

## Change checklist

- [x] Self-review.
- [x] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [x] Tests if relevant.

### CODE REVIEW

File: iroh/tests/patchbay.rs [hunk 1]
```diff
@@ -0,0 +1,1046 @@
...
+//! ```sh
+//! # On Linux (with user namespace support):
+//! RUSTFLAGS="--cfg patchbay_tests" cargo test --release -p iroh --test patchbay -- --test-threads=1
+//!
+//! # On macOS (via patchbay-vm):
+//! RUSTFLAGS="--cfg patchbay_tests" patchbay-vm test --release -p iroh --test patchbay -- --test-threads=1
+//! ```
+
+// patchbay only runs on linux
+#![cfg(target_os = "linux")]
```
Comment (dignifiedquire) [inline iroh/tests/patchbay.rs]: this seems to indicate it won't compile on macos with the command above?

File: iroh/tests/patchbay.rs [hunk 2]
```diff
@@ -0,0 +1,1046 @@
...
+                ping_open(&conn, timeout).await?;
+                Ok(())
+            },
+        )
+        .await?;
+    guard.ok();
+    Ok(())
+}
+
+// ---
```
Comment (dignifiedquire) [inline iroh/tests/patchbay.rs:374]: this is becoming a long file, maybe split it up into multiple test files?

File: iroh/tests/patchbay.rs [hunk 3]
```diff
@@ -0,0 +1,1046 @@
...
+        }
+    }
+    Ok((last_pass, guard))
+}
+
+/// Impaired side is the accepting (server) peer.
+#[tokio::test]
+#[traced_test]
+#[serial_test::serial]
+async fn degrade_ladder_impaired_server() -> Result {
```
Comment (dignifiedquire) [inline iroh/tests/patchbay.rs]: instead of this loop, can we not make each level its own?

File: iroh/tests/patchbay.rs (hunk 1 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: good question. the command above will cross-compile to linux, so it should work as expected.

I just merged https://github.com/n0-computer/patchbay/pull/7 so patchbay now uses  `container` on macos which is much faster than qemu.

Would be great if some people on mac could test this out, and see if it works with the instructions in the test module here.

**Comment by flub** [comment]
> cargo install --git https://github.com/n0-computer/patchbay -p patchbay-runner
> patchbay serve --testdir --open

Why the need to install a tool, can you not write a self-contained html file into the testdir?

File: iroh/Cargo.toml [hunk 4]
```diff
@@ -138,6 +141,12 @@ console = { version = "0.16" }
 wasm-tracing = "2.1.0"
 wasm-bindgen-test = "0.3.62"
 
+# patchbay netsim test dependencies (linux only)
+[target.'cfg(target_os = "linux")'.dev-dependencies]
+ctor = "0.6"
+patchbay = { git = "https://github.com/n0-computer/patchbay.git", branch = "main" }
+testdir = "0.9"
```
Comment (flub) [inline iroh/Cargo.toml]:
```suggestion
testdir = "0.10"
```
for some nextest fixes

File: iroh/tests/patchbay.rs [hunk 5]
```diff
@@ -0,0 +1,1046 @@
...
+use n0_error::{Result, StackResultExt, StdResultExt};
+use n0_tracing_test::traced_test;
+use patchbay::{Firewall, LinkCondition, LinkLimits, Nat, RouterPreset, TestGuard};
+use testdir::testdir;
+use tracing::info;
+
+use self::util::{Pair, PathWatcherExt, lab_with_relay, ping_accept, ping_open};
+
+#[path = "patchbay/util.rs"]
+mod util;
```
Comment (flub) [inline iroh/tests/patchbay.rs:41]: This is a very strange construct. You can anchor it to `self::util` or so, no? But even without this should work?

File: iroh/tests/patchbay.rs [hunk 6]
```diff
@@ -0,0 +1,1046 @@
...
+// NAT type matrix: verify holepunching across different NAT combinations
+// ---
+
+/// One peer behind Home NAT, the other on a public network.
+/// Holepunching should succeed: EIM mapping means the public peer can reach
+/// the NATted peer's mapped port once it learns the address via relay.
+#[tokio::test]
+#[traced_test]
+#[serial_test::serial]
+#[ignore = "stays relayed, holepunch times out (deadline elapsed)"]
```
Comment (flub) [inline iroh/tests/patchbay.rs]: TIL that `cargo test` now prints the ignore reasons, that's nice! Can't wait for nextest to catch up.

File: iroh/tests/patchbay.rs [hunk 7]
```diff
@@ -0,0 +1,1046 @@
...
+
+// ---
+// Holepunch tests
+// ---
+
+/// Simple holepunch: Two devices behind destination-independent NATs,
+/// establish via relay, upgrade to direct.
+#[tokio::test]
+#[traced_test]
+#[serial_test::serial]
```
Comment (flub) [inline iroh/tests/patchbay.rs]: we also have a mechanism in `.config/nextest.toml` that makes tests run in serial on nextest. How does this interact? Running this stuff in nextest right now results in a lot of failures, while nextest is our default test runner currently.

I guess from this it means you can't run tests concurrently? This kind of surprises me. Why is it not possible to give each test it's own namespace and let them run concurrently?

File: iroh/tests/patchbay.rs [hunk 8]
```diff
@@ -0,0 +1,1046 @@
...
+        .build()
+        .await?;
+    let dev1 = lab.add_device("dev1").uplink(public.id()).build().await?;
+    let dev2 = lab.add_device("dev2").uplink(home.id()).build().await?;
+    let timeout = Duration::from_secs(10);
+    Pair::new(dev1, dev2, relay_map)
+        .run(
+            async move |_dev, _ep, conn| {
+                let mut paths = conn.paths();
+                assert!(paths.is_relay(), "connection started relayed");
```
Comment (flub) [inline iroh/tests/patchbay.rs]: Why is this not flaky? This checks to see whether the selected path is a relay path, but this test does not itself control when holepunching starts so I would expect this to be flaky. If the runtime manages to delay this task enough then iroh could already have holepunched.

File: .github/workflows/patchbay.yml [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        run: sudo sysctl -w kernel.apparmor_restrict_unprivileged_userns=0
+        continue-on-error: true
+
+      - uses: actions/checkout@v6
+      - uses: dtolnay/rust-toolchain@stable
+      - uses: mozilla-actions/sccache-action@v0.0.9
+
+      - name: Run patchbay tests
+        id: tests
+        run: cargo test --release -p iroh --test patchbay -- --test-threads=1 --nocapture
```
Comment (matheus23) [inline .github/workflows/patchbay.yml]: Locally, I don't need `--test-threads=1` anymore.
Should we update this to `cargo nextest` instead?

```suggestion
        run: cargo nextest run --release -p iroh --test patchbay -P patchbay --no-fail-fast
```

Also needs an addition to `nextest.toml`:
```toml
[profile.patchbay]
slow-timeout = { period = "100s", terminate-after = 5 }
```
So it doesn't time out that quickly.

Not sure what to do about `--nocapture` in this case, though.

File: iroh/tests/patchbay.rs (hunk 7 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: I will remove the serial annotations and limit concurrency with nextest instead.

**Comment by matheus23** [comment]
The command in the PR description needs to be updated: The tests are now called `patchbay`, not `netsim`

**Comment by matheus23** [comment]
>     cargo install --git https://github.com/n0-computer/patchbay -p patchbay-runner
>     patchbay serve --testdir --open

This command from the PR description needs updating: It's called `patchbay-cli`, and you need to drop the `-p`, as that's not an allowed flag in `cargo install`.

File: iroh/tests/patchbay.rs (hunk 5 above)
Comment (matheus23) [reply iroh/tests/patchbay.rs:41]: I'd be in favor of doing this construction. Otherwise you need to move `util.rs` into the same directory as `patchbay.rs`, and it'll assume `util.rs` is another test binary, compiling it twice if you run `cargo test`. A little annoying.

File: iroh/tests/patchbay.rs (hunk 7 above)
Comment (matheus23) [reply iroh/tests/patchbay.rs]: `serial_test` is removed now, we don't need to run these tests serially. 
The dependency needs to removed from the `Cargo.toml` still, though.

File: iroh/tests/patchbay/util.rs [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+        f(&self.selected())
+    }
+
+    fn is_ip(&mut self) -> bool {
+        self.match_selected(PathInfo::is_ip)
+    }
+
+    fn is_relay(&mut self) -> bool {
+        self.match_selected(PathInfo::is_relay)
+    }
```
Comment (matheus23) [inline iroh/tests/patchbay/util.rs]: This is some very verbose code.
`is_ip` is the same as `self.selected().is_ip()`. No need to go via `match_selected`.
I'd say all of this can be removed.

The `wait_` stuff is useful though I can see that.

File: iroh/Cargo.toml [hunk 11]
```diff
@@ -101,10 +101,12 @@ getrandom = { version = "0.3.2", features = ["wasm_js"] }
 # target-common test/dev dependencies
 [dev-dependencies]
 console_error_panic_hook = "0.1"
+n0-error = { version = "0.1", features = ["anyhow"] }
 postcard = { version = "1.1.1", features = ["use-std"] }
 tracing-subscriber = { version = "0.3", features = ["env-filter"] }
 rand_chacha = "0.9"
 chrono = "0.4.43"
+serial_test = "3.4.0"
```
Comment (matheus23) [inline iroh/Cargo.toml]: We don't need this dependency anymore.

File: iroh/tests/patchbay.rs [hunk 12]
```diff
@@ -0,0 +1,999 @@
...
+//! RUSTFLAGS="--cfg patchbay_tests" cargo test --release -p iroh --test patchbay -- --test-threads=1
+//!
+//! # On macOS (via patchbay-vm):
+//! RUSTFLAGS="--cfg patchbay_tests" patchbay-vm test --release -p iroh --test patchbay -- --test-threads=1
+//! ```
+
+// patchbay only runs on linux
+#![cfg(target_os = "linux")]
+// Only compile these tests when the patchbay_tests cfg is enabled.
+#![cfg(patchbay_tests)]
```
Comment (matheus23) [inline iroh/tests/patchbay.rs]: RUSTFLAGS configs are really cumbersome to use.

I'd propose using nextest profiles instead. I've opened https://github.com/n0-computer/iroh/pull/4059 for this.

Comment (Frando) [reply iroh/tests/patchbay.rs]: Thanks, I had the same idea just now ;) merged your changes into my changes and pushed.

File: iroh/Cargo.toml (hunk 11 above)
Comment (Frando) [reply iroh/Cargo.toml]: Removed

File: iroh/tests/patchbay/util.rs (hunk 10 above)
Comment (Frando) [reply iroh/tests/patchbay/util.rs]: Yeah there's some leftovers from a previous iteration. Will clean up.

File: iroh/Cargo.toml (hunk 4 above)
Comment (Frando) [reply iroh/Cargo.toml]: Thanks, this is much better with nextest now.

File: iroh/tests/patchbay.rs (hunk 7 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: Removed the depencency too.

File: iroh/tests/patchbay.rs (hunk 1 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: So the cfg linux is definitely correct, patchbay only compiles on linux. The usage via `patchbay` cli on macos is unaffacted.

File: iroh/tests/patchbay.rs (hunk 3 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: Did this.

File: .github/workflows/patchbay.yml (hunk 9 above)
Comment (Frando) [reply .github/workflows/patchbay.yml]: The patchbay profile was added.

File: .config/nextest.toml [hunk 13]
```diff
@@ -15,4 +15,12 @@ test-group = 'run-in-isolation'
 threads-required = 32
 
 [profile.default]
+default-filter = 'not binary(patchbay)'
 slow-timeout = { period = "10s", terminate-after = 3 }
+
+[profile.patchbay]
+fail-fast = false
+retries = 1
```
Comment (matheus23) [inline .config/nextest.toml:23]: I'm not sure if we want `retries = 1`.
@dignifiedquire found that it's possible to get the tests to reliably never fail, even in "absurd" network conditions.
It might be a bad idea to enable this.

That said - perhaps we enable it for *this PR*, and once we've fixed issues, we remove the retries again.
Soooo let's file an issue, add that to a milestone and then make sure to remove it once stuff is stable?

Comment (Frando) [reply .config/nextest.toml:23]: sgtm!

Comment (Frando) [reply .config/nextest.toml:23]: Added an issue: https://github.com/n0-computer/iroh/issues/4061

**Comment by Frando** [comment]
Updated this to no longer use a git dependency (I just released patchbay 0.2.0)

File: iroh/tests/patchbay.rs (hunk 8 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: Right, this could be flaky if the runtime stalls this longer than at least one RTT. I've never seen it flake though. Should we leave it in until it flakes? So far it served me as a useful assert that the connection didn't start on IP, because if it did, a check for holepunching would be worthless.

**Comment by Frando** [comment]
> Why the need to install a tool, can you not write a self-contained html file into the testdir?

It would be possible, but when you double-click a html file it cannot `fetch` anything, so *everything* would have to be included in the HTML. Right now with `patchbay serve` logs and events are fetched on-demand.

I could build a version that allows to `http-serve ./taret/tesdir-current` and that would work (it would fetch things on demand) but then this would need more work and not sure if its any better than with `patchbay serve`.

File: iroh/tests/patchbay.rs (hunk 2 above)
Comment (Arqu) [reply iroh/tests/patchbay.rs:374]: +1

**Comment by Frando** [comment]
> One thing I'd like to know more about is how exactly the link degradation applies.

I digged into this, and actually discovered a bug in patchbay. The link conditioning was only applied on the device-to-router egress path, not on the ingress path. This is likely why the numbers didn't match up.

I have a PR for patchbay do fix this: https://github.com/n0-computer/patchbay/pull/12

I'll update this PR to use it for a round of testing, then merge and release patchbay, then update this PR back to the released version.

**Comment by matheus23** [comment]
> > One thing I'd like to know more about is how exactly the link degradation applies.
> 
> I digged into this, and actually discovered a bug in patchbay. The link conditioning was only applied on the device-to-router egress path, not on the ingress path. This is likely why the numbers didn't match up.

Yeah - this is what claude kept telling me, too, but I don't quite understand how that would explain the numbers. Any slowdown on the device-to-router egress path would put a delay on the whole round-trip, so how come a path can have an RTT lower than that?

**Comment by Frando** [comment]
>  Any slowdown on the device-to-router egress path would put a delay on the whole round-trip, so how come a path can have an RTT lower than that?

I don't really know. There's tests in patchbay that confirm the numbers with some variations, they seem to pass reliably. I'll try to test more myself.

**Comment by Frando** [comment]
I removed the "Publish to patchbay server" part for now. We don't want to hold this up on deploying the service, and the utility is debatable because we can just rerun locally if it fails.

So from my side this is ready for merge now. We can always iterate and add more tests later.

File: iroh/tests/patchbay.rs [hunk 14]
```diff
@@ -0,0 +1,1075 @@
...
+                Ok(())
+            },
+        )
+        .await?;
+    guard.ok();
+    Ok(())
+}
+
+/// Both peers behind CGNAT (EIM+EIF). The most permissive real-world NAT.
+/// Holepunching should succeed easily since filtering is endpoint-independent.
```
Comment (matheus23) [inline iroh/tests/patchbay.rs]: This claims that CGNAT is the most permissive NAT type?

File: iroh/tests/patchbay.rs [hunk 15]
```diff
@@ -0,0 +1,1075 @@
...
+            },
+        )
+        .await?;
+    guard.ok();
+    Ok(())
+}
+
+/// Both peers behind FullCone NAT (EIM+EIF with hairpin). The most permissive
+/// NAT type — any external host can send to the mapped port. Holepunching
+/// always succeeds on the first try.
```
Comment (matheus23) [inline iroh/tests/patchbay.rs]: And then this repeats the claim, but this time about FullCone NAT?

File: iroh/tests/patchbay.rs [hunk 16]
```diff
@@ -0,0 +1,1075 @@
...
+            async move |dev, _ep, conn| {
+                let mut paths = conn.paths();
+                paths.wait_ip(timeout).await.context("initial holepunch")?;
+                info!("holepunched, now killing link for 2s");
+
+                // Take the link down.
+                dev.link_down("eth0").await?;
+                tokio::time::sleep(Duration::from_secs(2)).await;
+                dev.link_up("eth0").await?;
+                info!("link restored, waiting for recovery");
```
Comment (matheus23) [inline iroh/tests/patchbay.rs]: Might be interesting to increase the time the link is down here - perhaps on the order of 15s or so - so that it's above the path idle timeout.

But I'm just musing. This is great already - we can add tests later.

File: iroh/tests/patchbay.rs [hunk 17]
```diff
@@ -0,0 +1,1075 @@
...
+            },
+        )
+        .await?;
+    guard.ok();
+    Ok(())
+}
+
+/// Hotel WiFi: captive-portal firewall allows all outbound TCP but only UDP
+/// port 53 (DNS). Similar to corporate firewall but less restrictive on TCP.
+/// Relay via HTTPS should work, holepunching should not.
```
Comment (matheus23) [inline iroh/tests/patchbay.rs]: Looking at the differences between `RouterPreset::Hotel` and `RouterPreset::Corporate`, the main difference seems to be that the hotel one is IPv4-only, whereas corporate is IPv4 + IPv6, and corporate only allows port 80/443, whereas the hotel one allows any port.

Tbqh, I don't think any of these things make much of a difference in terms of testing for iroh. In most cases I'd expect these tests to either both fail or none of them to fail, except maybe for some edge cases regarding IPv6.

Regardless, perhaps it's worth either adding more information to the test description, or linking the docs via something like "See also [`RouterPreset::Hotel`] and [`RouterPreset::Corporate`] for differences".

**Review by matheus23 — APPROVED** [review]
Just nits

File: .github/workflows/ci.yml [hunk 18]
```diff
@@ -31,6 +31,8 @@ jobs:
     if: "github.event_name != 'pull_request' || ! contains(github.event.pull_request.labels.*.name, 'flaky-test')"
     timeout-minutes: 30
     runs-on: [self-hosted, linux, X64]
+    env:
+      RUSTFLAGS: "-Dwarnings --cfg skip_patchbay"
```
Comment (flub) [inline .github/workflows/ci.yml:35]: is this still needed?

File: .github/workflows/patchbay.yml [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+  SCCACHE_CACHE_SIZE: "10G"
+  IROH_FORCE_STAGING_RELAYS: "1"
+  NEXTEST_VERSION: "0.9.132"
+
+jobs:
+  patchbay_tests:
+    name: Patchbay Tests
+    permissions:
+      contents: read
+      pull-requests: write
```
Comment (flub) [inline .github/workflows/patchbay.yml]: Does it currently need this? It seems weird to give permissions that are not needed.

File: .github/workflows/tests.yaml [hunk 20]
```diff
@@ -25,7 +25,7 @@ env:
   SCCACHE_CACHE_SIZE: "10G"
   CRATES_LIST: "iroh,iroh-bench,iroh-dns-server,iroh-relay"
   IROH_FORCE_STAGING_RELAYS: "1"
-  NEXTEST_VERSION: "0.9.80"
+  NEXTEST_VERSION: "0.9.132"
```
Comment (flub) [inline .github/workflows/tests.yaml]: Are you bumping this for fun or for a good reason? We had flaky behaviour with nextest at some point and that is why we pinned it. It might be fixed now, in which case I'm all for bumping it. But it would be better to do this in a separate PR if possible.

File: .gitignore [hunk 21]
```diff
@@ -1,3 +1,4 @@
 /target
 /logs
 iroh.config.toml
+/.patchbay
```
Comment (flub) [inline .gitignore:4]: what files do get stored here?

File: Cargo.toml [hunk 22]
```diff
@@ -36,7 +36,7 @@ missing_debug_implementations = "warn"
 # do.  To enable for a crate set `#![cfg_attr(iroh_docsrs,
 # feature(doc_cfg))]` in the crate.
 # We also have our own `iroh_loom` cfg to enable tokio-rs/loom testing.
-unexpected_cfgs = { level = "warn", check-cfg = ["cfg(iroh_docsrs)", "cfg(iroh_loom)"] }
+unexpected_cfgs = { level = "warn", check-cfg = ["cfg(iroh_docsrs)", "cfg(iroh_loom)", "cfg(skip_patchbay)"] }
```
Comment (flub) [inline Cargo.toml:39]: this is also no longer used?

File: Makefile.toml [hunk 23]
```diff
@@ -26,3 +26,8 @@ args = [
      "--config",
      "imports_granularity=Crate,group_imports=StdExternalCrate,reorder_imports=true,format_code_in_doc_comments=true",
 ]
+
+[tasks.patchbay]
+workspace = false
+command = "cargo"
+args = ["nextest", "run", "--profile", "patchbay"]
```
Comment (flub) [inline Makefile.toml]: :heart:

File: iroh/tests/patchbay/util.rs [hunk 24]
```diff
@@ -0,0 +1,359 @@
...
+        let task1 = self.dev1.spawn(move |dev| {
+            async move {
+                let endpoint = endpoint_builder(&dev, relay_map2).bind().await?;
+                endpoint.online().await;
+                addr_tx.send(addr_relay_only(endpoint.addr())).unwrap();
+                let conn = endpoint.accept().await.unwrap().accept().anyerr()?.await?;
+                watch_selected_path(&conn);
+                peer1(dev.clone(), endpoint.clone(), conn.clone()).await?;
+                conn.closed().await;
+                endpoint.close().await;
```
Comment (flub) [inline iroh/tests/patchbay/util.rs]: This (and the same for the peer) should have a few more `info!()` messages. I would add one least one before and after the callback.

File: iroh/tests/patchbay/util.rs [hunk 25]
```diff
@@ -0,0 +1,359 @@
...
+                Err(err) => Err(anyerr!(err, "device {} panicked", dev.name())),
+                Ok(Err(err)) => Err(anyerr!(err, "device {} failed", dev.name())),
+                Ok(Ok(())) => Ok(()),
+            };
+            let res_str = res.as_ref().map_err(|err| format!("{err:#}")).cloned();
+            dev.run_sync(move || {
+                match res_str {
+                    Ok(()) => {
+                        tracing::event!(
+                            target: "iroh::_events::test_ok",
```
Comment (flub) [inline iroh/tests/patchbay/util.rs]: Heh, I this is a bit abusing the events... it should be kind of reserved for iroh-the-library. I'm not yet sure how I feel about this.

 Can we make it all in the `iroh::_events::test::` namespace at least?

File: iroh/tests/patchbay/util.rs [hunk 26]
```diff
@@ -0,0 +1,359 @@
...
+                }
+                self.updated().await?;
+            }
+        })
+        .await
+        .anyerr()?
+    }
+}
+
+pub async fn ping_open(conn: &Connection, timeout: Duration) -> Result {
```
Comment (flub) [inline iroh/tests/patchbay/util.rs:297]: A succinct doc-comment would help reading tests with the help of rust-analyzer annotations. Much like you have them for wait_ip and wait_relay.

Also for accept.

File: iroh/tests/patchbay.rs [hunk 27]
```diff
@@ -0,0 +1,1075 @@
...
+//!
+//! These tests require Linux with user namespace support. On non-Linux systems, you can use
+//! the `patchbay` CLI to get a Linux container or VM with the required capabilities.
+//! See patchbay docs for details.
+//!
+//! To run:
+//!
+//! ```sh
+//! # On Linux (with user namespace support):
+//! cargo nextest run -p iroh --test patchbay -P patchbay
```
Comment (flub) [inline iroh/tests/patchbay.rs]: let's also mention cargo make here

File: iroh/tests/patchbay.rs (hunk 5 above)
Comment (flub) [reply iroh/tests/patchbay.rs:41]: huh, integration tests are build weirdly. Had to try this out to understand it.

File: iroh/tests/patchbay.rs [hunk 28]
```diff
@@ -0,0 +1,1075 @@
...
+        .await?;
+    guard.ok();
+    Ok(())
+}
+
+/// Tests that changing the uplink of an interface works (i.e. switching wifis).
+///
+/// For this we observe a change in the selected path's remote addr on the *other* side.
+/// Whether the side that changes interfaces opens a new path or does an RFC9000-style migration
+/// is an implementation detail which we won't test for.
```
Comment (flub) [inline iroh/tests/patchbay.rs]: Not sure this makes sense, because if it were to do an RFC9000-style migration the other side would not see a change in the selected path either I think.

File: iroh/tests/patchbay/util.rs [hunk 29]
```diff
@@ -0,0 +1,359 @@
...
+    /// [`EndpointAddr`], and then the supplied functions are invoked, passing
+    /// the device, endpoint, and connection to user code.
+    ///
+    /// After a future complete, `peer1` awaits the connection to be closed,
+    /// whereas `peer2` closes the connection.
+    ///
+    /// Afterwards, both endpoints are closed and metrics are recorded through
+    /// [`Device::record_iroh_metrics`]. Will also emit a debug log with target
+    /// `patchbay::_events` with the result of the user-supplied work functions.
+    pub async fn run<F1, Fut1, F2, Fut2>(self, peer1: F1, peer2: F2) -> Result
```
Comment (flub) [inline iroh/tests/patchbay/util.rs]: One issue that I have with this signature is that it is really hard in a test to tell which one of the two initiates the connection and which one accepts it (I'm in a review, don't have rust-analyzer here). It is kind of made worse by being disconnected from the new signature as well where the same order matters.

iroh might be p2p, but QUIC has definite server and client roles. And understanding those is important in tests.

I don't have a clever solution just yet. And also don't want to block on just this, but maybe someone else has a good idea.

File: iroh/tests/patchbay.rs [hunk 30]
```diff
@@ -0,0 +1,1075 @@
...
+// NAT type matrix: verify holepunching across different NAT combinations
+// ---
+
+/// One peer behind Home NAT, the other on a public network.
+/// Holepunching should succeed: EIM mapping means the public peer can reach
+/// the NATted peer's mapped port once it learns the address via relay.
+#[tokio::test]
+#[traced_test]
+#[ignore = "stays relayed, holepunch times out (deadline elapsed)"]
+async fn holepunch_home_nat_one_side() -> Result {
```
Comment (flub) [inline iroh/tests/patchbay.rs]: it is strange this doens't work. what mechanism did we figure to address these ignored tests?

**Comment by flub** [comment]
Oh, another thing I found was that investigating any failure requires me to set RUST_LOG=trace. Why not enable that by default for the logfiles created? It uses some more disk space, but I don't think that's a big deal. Any other downside?

File: Cargo.toml (hunk 22 above)
Comment (Frando) [reply Cargo.toml:39]: It is, see above. It's used in the cross tests to skip running the patchbay tests.

File: iroh/tests/patchbay/util.rs (hunk 25 above)
Comment (Frando) [reply iroh/tests/patchbay/util.rs]: Why abusing?
Can change the namespace, sure. It's really helpful to have this though because this makes this show up in the patchbay timeline view in the web ui. I wrote that one such that it shows all tracing lines with `_events`  in the target. So sure, I can change it to `iroh::_events::test`.

**Comment by Frando** [comment]
> there are too many tests to review, i'm not sure how to tackle that. 

I could just remove some, and do another PR then to add more. Let me do that tomorrow.

File: iroh/tests/patchbay.rs (hunk 30 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: No clear mechanism yet. First step: Let's run them locally with `--ignored` and if someone has an ignored one passing, then do a PR to remove the ignore line and we see what CI says.

File: iroh/tests/patchbay/util.rs (hunk 29 above)
Comment (Frando) [reply iroh/tests/patchbay/util.rs]: `peer1` is server, `peer2` is client. I'll rename the args. And think a bit if I find a better way to express this. I do quite like how simple it is to use this in the tests though.

File: .github/workflows/ci.yml (hunk 18 above)
Comment (matheus23) [reply .github/workflows/ci.yml:35]: Yes, this skips patchbay tests for the cross-compilation job. This is needed. You might be mistaking this cfg with the previous iteration that had the `patchbay_test` cfg. This one is different.

File: iroh/tests/patchbay.rs (hunk 14 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: This test is among the ones I removed for the inital merge. I will review the tests and docs more and mark https://github.com/n0-computer/iroh/pull/4065 as ready for review once one.

File: iroh/tests/patchbay.rs (hunk 16 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: Changed to 5s. Can adapt further if needed, or we add more tests with that.

File: iroh/tests/patchbay.rs (hunk 15 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: This test is among the ones I removed for the inital merge. I will review the tests and docs more and mark https://github.com/n0-computer/iroh/pull/4065 as ready for review once one.

File: iroh/tests/patchbay.rs (hunk 17 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: :thumbsup: This test is among the ones I removed for the inital merge. I will review the tests and docs more and mark https://github.com/n0-computer/iroh/pull/4065 as ready for review once one.

File: .github/workflows/tests.yaml (hunk 20 above)
Comment (Frando) [reply .github/workflows/tests.yaml]: At one point I used an option that was only supported in the later version. I don't think this is the case though, so will downgrade again.

File: .gitignore (hunk 21 above)
Comment (Frando) [reply .gitignore:4]: If you use the `patchbay` cli this is where test results will be stored.

File: iroh/tests/patchbay/util.rs (hunk 24 above)
Comment (Frando) [reply iroh/tests/patchbay/util.rs]: I added more info messages.

File: iroh/tests/patchbay/util.rs (hunk 26 above)
Comment (Frando) [reply iroh/tests/patchbay/util.rs:297]: Added comments.

File: iroh/tests/patchbay.rs (hunk 27 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: Fixed

File: iroh/tests/patchbay.rs (hunk 28 above)
Comment (Frando) [reply iroh/tests/patchbay.rs]: Reworded the comment

File: iroh/tests/patchbay/util.rs (hunk 29 above)
Comment (Frando) [reply iroh/tests/patchbay/util.rs]: I rewrote this to have clearly marked `server` and `client` functions.

**Comment by Frando** [comment]
I moved a number of tests to https://github.com/n0-computer/iroh/pull/4065 to make it easier to review this PR.
Also did a round of cleanups, I think I addressed all feedback.

File: .github/workflows/patchbay.yml [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+
+concurrency:
+  group: patchbay-${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+env:
+  RUST_BACKTRACE: 1
+  SCCACHE_CACHE_SIZE: "10G"
+  IROH_FORCE_STAGING_RELAYS: "1"
+  NEXTEST_VERSION: "0.9.132"
```
Comment (matheus23) [inline .github/workflows/patchbay.yml]: The nextest version we have here is still 0.9.132, perhaps we should align the versions to match, if possible (esp. since we haven't verified that 0.9.132 doesn't have the same bug that versions after 0.9.80 had).

**Review by matheus23 — APPROVED** [review]

File: iroh/tests/patchbay.rs (hunk 2 above)
Comment (Frando) [reply iroh/tests/patchbay.rs:374]: Removed several tests and moved them to https://github.com/n0-computer/iroh/pull/4065 (still a draft) so let's do this once we add those back.

File: .github/workflows/patchbay.yml (hunk 31 above)
Comment (Frando) [reply .github/workflows/patchbay.yml]: changed to 0.9.80

File: iroh/tests/patchbay/util.rs (hunk 25 above)
Comment (flub) [reply iroh/tests/patchbay/util.rs]: Abusing because these are not events emitted by iroh. They are emitted by tests. They are only emitted in the same way because patchbay already had code that was looking for these events. I don't think that's the cleanest design. It's a bit like emitting events in another libraries logging targets.

But also I'm fine with for now containing this into a specific namespace that hopefully won't naturally occur in iroh itself. So let's not spend too long on this.

File: iroh/tests/patchbay/util.rs (hunk 29 above)
Comment (flub) [reply iroh/tests/patchbay/util.rs]: oh, that's rather nice! :heart:
